### PR TITLE
Add support for AGP 8.4

### PR DESCRIPTION
**Pull request by emersion**
Fixes the following errors with AGP 8.4:

    A problem occurred configuring project ':webcrypto'.
    > [CXX1110] Platform version 16 is unsupported by this NDK. Please change minSdk to at least 21 to avoid undefined behavior. To suppress this error, add android.ndk.suppressMinSdkVersionError=21 to the project's gradle.properties or set android.experimentalProperties["android.ndk.suppressMinSdkVersionError"]=21 in the Gradle build file.

and

    Execution failed for task ':webcrypto:processReleaseManifest'.
    > A failure occurred while executing com.android.build.gradle.tasks.ProcessLibraryManifest$ProcessLibWorkAction
       > Incorrect package="com.example.webcrypto" found in source AndroidManifest.xml: /home/simon/src/webcrypto.dart/android/src/main/AndroidManifest.xml.
         Setting the namespace via the package attribute in the source AndroidManifest.xml is no longer supported.
         Recommendation: remove package="com.example.webcrypto" from the source AndroidManifest.xml: /home/simon/src/webcrypto.dart/android/src/main/AndroidManifest.xml.

### CODE REVIEW

File: android/build.gradle [hunk 1]
```diff
@@ -53,7 +50,7 @@ android {
     }
 
     defaultConfig {
-        minSdkVersion 16
+        minSdkVersion 21
```
Comment (emersion) [inline android/build.gradle:53]: I wonder if we should just use `flutter.minSdkVersion` somehow? Did not seem to work in my testing though:

```
> Could not get unknown property 'flutter' for extension 'android' of type com.android.build.gradle.LibraryExtension.
```

File: android/build.gradle [hunk 2]
```diff
@@ -25,10 +25,7 @@ apply plugin: 'com.android.library'
 apply plugin: 'kotlin-android'
 
 android {
-    // Conditional for compatibility with AGP <4.2.
-    if (project.android.hasProperty("namespace")) {
-        namespace 'dev.google.webcrypto'
-    }
+    namespace 'dev.google.webcrypto'
 
     compileSdkVersion 31
```
Comment (emersion) [inline android/build.gradle:30]: Is there a reason for this being 31 instead of `flutter.ndkVersion`? (Currently 23.1.7779620)

**Comment by jonasfj** [comment]
Do I understand it correctly that this will drop support for Android 4.x, which is more than 12 years old at this point.

That might be acceptable, I'm guessing :D

File: android/build.gradle [hunk 3]
```diff
@@ -25,10 +25,7 @@ apply plugin: 'com.android.library'
 apply plugin: 'kotlin-android'
 
 android {
-    // Conditional for compatibility with AGP <4.2.
-    if (project.android.hasProperty("namespace")) {
-        namespace 'dev.google.webcrypto'
-    }
+    namespace 'dev.google.webcrypto'
```
Comment (jonasfj) [inline android/build.gradle:28]: Dumb question, but do we not need this anymore?

Will `minSdkVersion 21` make AGP < 4.2 irrelevant?

**Comment by jonasfj** [comment]
Hmm, I'm guessing we should also look at:
https://docs.flutter.dev/release/breaking-changes/flutter-gradle-plugin-apply

I'll file it separately, as https://github.com/google/webcrypto.dart/issues/124

**Review by jonasfj — APPROVED** [review]
Is there some kind of testing we should do to be more confident this is correct?

I'd love to have it covered in CI.

**Comment by HamdaanAliQuatil** [comment]
> Fixes the following errors with AGP 8.4:
> 
> ```
> A problem occurred configuring project ':webcrypto'.
> > [CXX1110] Platform version 16 is unsupported by this NDK. Please change minSdk to at least 21 to avoid undefined behavior. To suppress this error, add android.ndk.suppressMinSdkVersionError=21 to the project's gradle.properties or set android.experimentalProperties["android.ndk.suppressMinSdkVersionError"]=21 in the Gradle build file.
> ```
> 
> and
> 
> ```
> Execution failed for task ':webcrypto:processReleaseManifest'.
> > A failure occurred while executing com.android.build.gradle.tasks.ProcessLibraryManifest$ProcessLibWorkAction
>    > Incorrect package="com.example.webcrypto" found in source AndroidManifest.xml: /home/simon/src/webcrypto.dart/android/src/main/AndroidManifest.xml.
>      Setting the namespace via the package attribute in the source AndroidManifest.xml is no longer supported.
>      Recommendation: remove package="com.example.webcrypto" from the source AndroidManifest.xml: /home/simon/src/webcrypto.dart/android/src/main/AndroidManifest.xml.
> ```

@emersion can you please let me know how to reproduce this error

File: android/build.gradle (hunk 1 above)
Comment (emersion) [reply android/build.gradle:53]: Seems like plugins can't do that.

**Comment by emersion** [comment]
> Do I understand it correctly that this will drop support for Android 4.x, which is more than 12 years old at this point.

Correct. Note that the latest version of Flutter [requires Android ≥ 5.x](https://docs.flutter.dev/reference/supported-platforms), so this just aligns the plugin requirements with Flutter's.

> Hmm, I'm guessing we should also look at:
https://docs.flutter.dev/release/breaking-changes/flutter-gradle-plugin-apply

Yes, that would be nice to fix as well. But it's orthogonal.

> Is there some kind of testing we should do to be more confident this is correct?
>
> I'd love to have it covered in CI.

I've pushed a commit to upgrade the example to AGP 8.4. The failures reported in the first two commits can be reproduced there.

File: android/build.gradle (hunk 3 above)
Comment (emersion) [reply android/build.gradle:28]: AGP 8.4 rejects `AndroidManifest.xml` files with a `package` attribute. If we only set the namespace for AGP ≥ 4.2, then on AGP < 4.2 we won't define any namespace at all and the build will fail anyways.

I think it's reasonable to ask library users to upgrade AGP ≥ 4.2 when upgrading webcrypto.dart.

**Pull request merged**